src/libDirectoryService/DSBlockPreProcessing.cpp

```diff
@@ -1216,6 +1366,124 @@ bool DirectoryService::ProcessShardingStructure(
+            LOG_GENERAL(WARNING,
+                        "Unknown (Not in DS Committee) public key "
+                            << pubkey
+                            << " found to have "
+                               "contributed co-sigs as a DS Committee member.");
+          } else {
+            // Increment the performance score if the public key exists.
+            ++m_dsMemberPerformance[pubkey];
+          }
+        }
+      }
+    }
+  }
+
+  // Display the performance scores of all the DS Committee members.
+  LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+            "DS Committee Co-Signature Performance");
+  unsigned int index = 0;
+  uint32_t maxCoSigs = (NUM_FINAL_BLOCK_PER_POW - 1) * 2;
+  for (const auto& member : m_dsMemberPerformance) {
+    LOG_GENERAL(INFO, "[" << PAD(index++, 3, ' ') << "] " << member.first << " "
+                          << PAD(member.second, 4, ' ') << "/" << maxCoSigs);
+  }
+}
+
+unsigned int DirectoryService::InjectByzantineNodes(
+    unsigned int numOfProposedDSMembers, std::map<PubKey, Peer>& powDSWinners) {
+  LOG_MARKER();
+  std::lock_guard<mutex> g(m_mutexDsMemberPerformance);
+
+  // Do not inject Byzantine nodes on the first epoch when performance cannot be
+  // measured.
+  if (m_mediator.m_currentEpochNum <= 1) {
+    LOG_GENERAL(INFO,
+                "Skipping injecting Byzantine nodes for removal since "
+                "performance cannot be measured on the first epoch.");
+    return 0;
+  }
+
+  // Parameters
+  uint32_t maxCoSigs = (NUM_FINAL_BLOCK_PER_POW - 1) * 2;
+  uint32_t threshold = std::ceil(DS_PERFORMANCE_THRESHOLD_PERCENT * maxCoSigs);
+  unsigned int numToRemove =
+      std::min(NUM_DS_BYZANTINE_REMOVED, numOfProposedDSMembers);
+  unsigned int index;
+
+  // Build a list of Byzantine Nodes
+  LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+            "Evaluating performance of the current DS Committee.");
+  LOG_GENERAL(INFO, "maxCoSigs = " << maxCoSigs);
+  LOG_GENERAL(INFO, "threshold = " << threshold << " ("
+                                   << DS_PERFORMANCE_THRESHOLD_PERCENT << ")");
+  unsigned int numByzantine = 0;
+  index = 0;
+  for (auto it = m_mediator.m_DSCommittee->begin();
+       it != m_mediator.m_DSCommittee->end(); ++it) {
+    // Do not evaluate guard nodes.
+    if (GUARD_MODE && Guard::GetInstance().IsNodeInDSGuardList(it->first)) {
+      continue;
+    }
+
+    // Check if the score is below the calculated threshold.
+    uint32_t score = m_dsMemberPerformance.at(it->first);
+    if (score < threshold) {
+      // Only add the node to be removed if there is still capacity.
+      if (numByzantine < numToRemove) {
+        // Check if the public key that was found to a Byzantine node is mine
+        // because the network information for mine is zeroed out.
+        if (it->first == m_mediator.m_selfKey.second) {
+          powDSWinners[it->first] = Peer();
+        } else {
+          powDSWinners[it->first] = it->second;
+        }
```

Comment:
The under performanced DS node should kick out, if put it into powDSWinner, it will come back again?